File: String/longest-substring/longest_substring.cpp
#include <string>
#include <set>

using namespace std;


int length_of_longest_substring(string s)
{

    int max_length, i, j = 0;

    set<char> unique_char;

    while (i < s.length())
    {
        if (unique_char.find(s[i]) != unique_char.end())
        {
            unique_char.erase(s[j]);
            j++;
        }
        else
        {
            unique_char.insert(s[i]);
            i++;
            max_length = max(max_length, i - j);
        }
    }

    return max_length;
}

File: String/longest-substring/longest_substring.py
# Longest Substring Without Repeating Characters

# Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/

def length_of_longest_substring(s: str) -> int:
    max_length = 0
    unique_characters = set()

    i = j = 0

    while (i < len(s)):

        if s[i] in unique_characters:
            unique_characters.remove(s[j])
            j += 1
        else:
            unique_characters.add(s[i])
            i += 1
            max_length = max(max_length, i-j)

    return max_length
